modules/planning/reference_line/cos_theta_problem_interface.cc

```diff
@@ -555,5 +619,196 @@ void CosThetaProbleminterface::hessian_strcuture() {
+    //               (init_points_[i + 2].y() - init_points_[i + 1].y())));
+
+    // Denominator numerically protected
+    // *obj_value +=
+    //     weight_cos_included_angle_ *
+    //     acos(
+    //         (((x[mindex] - x[findex]) * (x[lindex] - x[mindex])) +
+    //          ((x[mindex + 1] - x[findex + 1]) *
+    //           (x[lindex + 1] - x[mindex + 1]))) /
+    //         (1e-8 + sqrt((x[mindex] - x[findex]) * (x[mindex] -
+    //         x[findex]) +
+    //                      (x[mindex + 1] - x[findex + 1]) *
+    //                          (x[mindex + 1] - x[findex + 1])) *
+    //                     sqrt((x[lindex] - x[mindex]) * (x[lindex] -
+    //                     x[mindex]) +
+    //                          (x[lindex + 1] - x[mindex + 1]) *
+    //                              (x[lindex + 1] - x[mindex + 1]))));
+
+    // InnerProd
+    // *obj_value -=
+    //     weight_cos_included_angle_ *
+    //     (((x[mindex] - x[findex]) * (x[lindex] - x[mindex])) +
+    //      ((x[mindex + 1] - x[findex + 1]) * (x[lindex + 1] - x[mindex +
+    //      1])));
+
+    // CrossProd
+    // *obj_value += weight_cos_included_angle_ *
+    //              (((x[findex] - x[mindex]) * (x[lindex + 1] - x[mindex +
+    //              1]))
+    //              -
+    //               ((x[lindex] - x[mindex]) * (x[findex + 1] - x[mindex +
+    //               1]))) *
+    //              (((x[findex] - x[mindex]) * (x[lindex + 1] - x[mindex +
+    //              1]))
+    //              -
+    //               ((x[lindex] - x[mindex]) * (x[findex + 1] - x[mindex +
+    //               1])));
+
+    // Midpoint distance
+    *obj_value += weight_cos_included_angle_ *
+                  (((x[findex] + x[lindex]) / 2.0 - x[mindex]) *
+                       ((x[findex] + x[lindex]) / 2.0 - x[mindex]) +
+                   ((x[findex + 1] + x[lindex + 1]) / 2.0 - x[mindex + 1]) *
+                       ((x[findex + 1] + x[lindex + 1]) / 2.0 - x[mindex + 1]));
+  }
+
+  // Total length
+  for (size_t i = 0; i < num_of_points_ - 1; ++i) {
+    size_t findex = i << 1;
+    size_t nindex = findex + 2;
+    *obj_value +=
+        weight_length_ *
+        ((x[findex] - x[nindex]) * (x[findex] - x[nindex]) +
+         (x[findex + 1] - x[nindex + 1]) * (x[findex + 1] - x[nindex + 1]));
+  }
+  return true;
+}
+
+/** Template to compute contraints */
+template <class T>
+bool CosThetaProbleminterface::eval_constraints(int n, const T* x, int m,
+                                                T* g) {
+  // fill in the positional deviation constraints
+  for (size_t i = 0; i < num_of_points_; ++i) {
+    size_t index = i << 1;
+    g[index] = x[index];
+    g[index + 1] = x[index + 1];
+  }
+  return true;
+}
+
+/** Method to generate the required tapes */
+void CosThetaProbleminterface::generate_tapes(int n, int m, int* nnz_jac_g,
+                                              int* nnz_h_lag) {
+  Ipopt::Number* xp = new double[n];
```

Comment:
Avoid old-school memory management.
Use `vector<double> xp;`, and get the raw pointer as `&xp[0]` if the upstream API requires.

Reply:
Part of them fixed. Some of them are not fixed successfully yet, will look into it afterwards